Making Grids for SectionLists

andersonaddo opened the pull request:
Hello @saleel !
I liked the simplicity of your design, and I wanted to add some functionality!
I made another class to make Grid [SectionLists](https://facebook.github.io/react-native/docs/sectionlist.html) that scroll vertically.
Instead of splitting the data list into smaller arrays using the `chunkArray` function only once, I do it several times for the different sections of the section list.

I think I'm almost there, but I'm currently getting an error for my test data:
```Encountered two children with the same key, `0:undefined`. ```

Maybe we could both look into finishing this! It would be helpful for many developers since its so simple!
I added in some temporary console logs for debugging.

saleel commented:
This looks great. I think that error is from the data as well.

I don't have much time right now to look through the code, but I will check and merge when it is stable. I think we can refactor and share many common code in both classes. 

We will also have to see how we can make both options available to the user.

andersonaddo commented:
Nice to see that you've responded so soon!

> I think we can refactor and share many common code in both classes.

The new class that I made is actually just a modified version of the class in the `index.js` file. So most of the code is already based off the original code.

> We will also have to see how we can make both options available to the user.

I made the `index.js` file import and export the new class that I made, so the user can simply just use it by importing from the index file like 
```import {SuperGridSectionList} from 'react-native-super-grid'```

It's unfortunate that you are so busy at the moment; hopefully you'll be able to review the code some time soon! In the meantime, I'll see if I can get the code working.

saleel commented:
Great.

I would be able to review and add the PR once you submit final version.

andersonaddo commented:
I think it's ready now, waiting for you to test and merge!

saleel commented:
Great, will test it soon and merge.

saleel commented:
Hi,

`import SuperGridSectionList from './GridSectionList'` should be `import SuperGridSectionList from './SuperGridSectionList'`

Please fix this.

Also, what the structure of items and renderItem props? Is it same as native SectionList? If not, please make it so.

We are also planning to make the signature of items and renderItem same as FlatList for the normal Grid class as well.

andersonaddo commented:
I fixed the import statement.
I can confirm that the prop structure for the items prop is identical to the sections prop for the SectionList, and I also know that the renderItems prop is nearly if not completely identical too.

I'll double check the renderItems prop, but it should be good.

saleel commented:
I just tested and managed to get it work. Basically, everything is working great.

But the renderItem signature is different from SectionList. In SectionList its like this
`renderItem={({ item, index, section }) => <Text key={index}>{item}</Text>}`

there is only one parameter, and it is an object with keys item, index and section. Can you make this change please?

In renderSectionHeader, its correct as you are directly passing the function to the SectionList.

Also, I would suggest to rename the items prop to sections, to match with SectionList again.

Another small this I noticed is, `let itemCopy = JSON.parse(JSON.stringify(items));`  can be changed to  `let itemCopy = [...items];`

I haven't gone through all code. But if we get it to work, I can merge and then we both can continue improving it.

I am sorry for asking you to make so many changes, but lets do this properly. Its a great feature :-)

andersonaddo commented:
I fixed the renderItem prop, and changed the items prop to sections.
I wasn't able to successfully change the `let itemCopy = JSON.parse(JSON.stringify(items));` to `let itemCopy = [...items];` since spread was only doing a shallow copy of the objects inside the items array.

andersonaddo commented:
Great, thanks!

andersonaddo commented:
Great, thanks!
